Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +25,75 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHours_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return (currentNew_ < 0);
+}
+
+int BatteryStateOfChargeService::getRemainingHours() const
+{
+    return remainingHours_;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return timeTillChargeOrDepletion_;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    QTime timeOld = timeNew_;
+    double currentOld = currentNew_;
+
+    currentNew_ = batteryData.current;
+    timeNew_ = batteryData.time;
+
+    if(!timeOld.isNull())
+    {
+        double avgCurrent = (currentOld + currentNew_) / 2;
+        ampHours_ += (avgCurrent * timeOld.secsTo(timeNew_) / SECONDS_TO_HOURS);
+    }
+
+    //calculate remaining time for charge or depletion
+    double totalHours;
+    int hours;
+    int minutes;
+    int seconds;
+    int milliseconds;
```

Comment:
These aren't used until after the if statement, so we can move it down.

I think we can also just declare and set these variables, ie instead of 
```
int hours
...
hours = ....
```
Just do
```
int hours = ...
```